src/main/java/org/truffleruby/core/module/ModuleOperations.java

```diff
@@ -536,31 +547,29 @@ private static MethodLookupResult lookupSuperMethod(DynamicObject declaringModul
             }
         }
 
+        // Nothing found
         return new MethodLookupResult(null, toArray(assumptions));
     }
 
-    @TruffleBoundary
-    private static MethodLookupResult lookupSuperMethodNoRefinements(DynamicObject declaringModule, String name,
-            DynamicObject objectMetaClass) {
-        assert RubyGuards.isRubyModule(declaringModule);
-        assert RubyGuards.isRubyModule(objectMetaClass);
+    private static DynamicObject[] getRefinementsFor(DeclarationContext declarationContext, DynamicObject module) {
+        if (declarationContext == null) {
+            return null;
+        }
+        return declarationContext.getRefinementsFor(module);
+    }
 
-        final ArrayList<Assumption> assumptions = new ArrayList<>();
-        boolean foundDeclaringModule = false;
-        for (DynamicObject module : Layouts.MODULE.getFields(objectMetaClass).ancestors()) {
-            if (module == declaringModule) {
-                foundDeclaringModule = true;
-            } else if (foundDeclaringModule) {
-                final ModuleFields fields = Layouts.MODULE.getFields(module);
-                assumptions.add(fields.getMethodsUnmodifiedAssumption());
-                final InternalMethod method = fields.getMethod(name);
-                if (method != null) {
-                    return new MethodLookupResult(method, toArray(assumptions));
-                }
-            }
+    private static DynamicObject getLookupTo(DynamicObject[] refinements, DynamicObject refinedModule, int i) {
+        // If we have more then one active refinement for C (where C is refined module):
+        //     R1.ancestors = [R1, A, C, ...]
+        //     R2.ancestors = [R2, B, C, ...]
+        //     R3.ancestors = [R3, D, C, ...]
+        // we should look among C ancestors only for the last active refinement:
+        // R3 -> D -> R2 -> B -> R1 -> A -> C -> ...
+        if (i != refinements.length - 1) {
```

Comment:
Do we ever need to look behind C? Or could we always just stop at C when searching refined methods?

Reply:
The brilliant idea: https://github.com/oracle/truffleruby/pull/2033/commits/ec393db77477cd8e6911d9b8fb45e6b995ae58de